### PR TITLE
Create module for generating bsp4rs

### PR DESCRIPTION
**Pull request by KatKlo**

### CODE REVIEW

File: codegen/src/main/kotlin/org/jetbrains/bsp/generators/bsp4rs/DeriveRenderer.kt [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        is Def.Service -> emptySet()
+        is Def.ClosedEnum<*> -> when (def.enumType) {
+            is EnumType.IntEnum -> setOf(DeriveOption.ENUM_INT_SERIALIZE)
+            is EnumType.StringEnum -> setOf(DeriveOption.ENUM_STR_SERIALIZE)
+        }
+
+        else -> setOf(DeriveOption.STANDARD_SERIALIZE)
+    }
+
+    private fun defToBlackList(def: Def): Set<DeriveOption> {
```
Comment (agluszak) [inline codegen/src/main/kotlin/org/jetbrains/bsp/generators/bsp4rs/DeriveRenderer.kt]: https://en.wikipedia.org/wiki/Blacklist_(computing)#Controversy_over_use_of_the_term

File: codegen/src/main/kotlin/org/jetbrains/bsp/generators/bsp4rs/DeriveRenderer.kt [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            DeriveOption.HASH,
+            DeriveOption.ORD,
+        )
+        derivesSet = derivesSet.plus(defToSerializeList(def))
+        derivesSet = derivesSet.minus(defToBlackList(def).toSet())
+
+        return this.render()
+    }
+
+    private fun defToSerializeList(def: Def) = when (def) {
```
Comment (agluszak) [inline codegen/src/main/kotlin/org/jetbrains/bsp/generators/bsp4rs/DeriveRenderer.kt]: why no return type?

File: codegen/src/main/kotlin/org/jetbrains/bsp/generators/bsp4rs/SerializationRenderer.kt [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    private fun prepareSerdeSet(hints: List<Hint>) {
+        val rename = hints.find { it is Hint.JsonRename }
+
+        return if (rename is Hint.JsonRename)
+            serdeSet = setOf(SerdeOption.Rename(rename.name))
+        else
+            serdeSet = emptySet()
+    }
+
+    private fun defToSerdeList(def: Def, untagged: Boolean) = when (def) {
```
Comment (agluszak) [inline codegen/src/main/kotlin/org/jetbrains/bsp/generators/bsp4rs/SerializationRenderer.kt]: please add return types everywhere where it's not unit

File: codegen/src/test/kotlin/Bsp4rsTest.kt [hunk 4]
```diff
@@ -0,0 +1,180 @@
+import org.jetbrains.bsp.generators.bsp4rs.RustRenderer
+import org.jetbrains.bsp.generators.ir.Hint
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+
+class Bsp4rsTest {
+    private val renderer = RustRenderer("", emptyList(), "")
+
+//    @Test
```
Comment (agluszak) [inline codegen/src/test/kotlin/Bsp4rsTest.kt]: why commented out?

File: codegen/src/main/kotlin/org/jetbrains/bsp/generators/bsp4rs/DeriveRenderer.kt (hunk 1 above)
Comment (KatKlo) [reply codegen/src/main/kotlin/org/jetbrains/bsp/generators/bsp4rs/DeriveRenderer.kt]: corrected

File: codegen/src/main/kotlin/org/jetbrains/bsp/generators/bsp4rs/SerializationRenderer.kt (hunk 3 above)
Comment (KatKlo) [reply codegen/src/main/kotlin/org/jetbrains/bsp/generators/bsp4rs/SerializationRenderer.kt]: done

File: codegen/src/test/kotlin/Bsp4rsTest.kt (hunk 4 above)
Comment (KatKlo) [reply codegen/src/test/kotlin/Bsp4rsTest.kt]: I forgot about this; fixed now

**Review by agluszak — APPROVED** [review]

**Pull request merged**